### PR TITLE
Fairy v0.6

### LINKED ISSUE #64: [0.6] Remove Hot Loading and made Fairy work per plugin
Currently fairy uses hot loading to load libraries and modules
And fairy core is an independent core that doesn't rely on any other plugins
Although it sounds fantastic, but in the end it is very promblematic.

## Reasons to remove hot loading:
- Java 16+ by default closes the possibility to access some packages in jvm, which also closed the possibility to access `URLClassLoader#addURL()`. This is the most critical issue.
- Due to how hot loading works, whenever you added a dependency that's built with fairy, you are in a risk to saw missing classes. such as the dependency has sidebar as their depended module, you wouldn't be able to access SidebarAdapter. (Even if you manually add sidebar as dependency, since every project relocates module, you still cannot access to it)

## Reasons to made fairy work per plugin:
- Versioning is way to complex at the moment. 
  - Fairy platform jar at the moment does not automatically updated and it's independent core, there is currently no workaround to ensure modules version and platform version. if the module uses newer core function, it will break.
  - Even if it's automatically updated, if plugins or modules using outdated functions, it will break.
  - Making workaround to ensure versions work together is way too complex.

## Advantages:
- Java 16+ support.
- Less plugin conflict.
- Plugin will be able to specify core version that this plugin should use. Making plugin more maintainable.
- Shading modules and core into plugin can ensure whenever other plugin depend on this plugin will not having missing classes.

## Disadvantage:
- Uber jar.
- Multiple plugins development will require more workaround.

## Works to do:
- [x] Work on bootstrap system, library loader system and gradle plugin to made them work with new design.
- [x] Reduce unnecessary libraries to ensure the final uber jar wouldn't be that huge.
- [ ] Made module memories current core version, reject plugin from using the modules whenever core version is out-dated.

### LINKED ISSUE #71: Made ComponentRegistry capable of handling multiples
At the moment each component are only being applied with 1 ComponentRegistry
It will restricted because ComponentRegistry handles constructor
But it is so unnecessary and useless, so rather remove it for much better flexibility for current component usage

### LINKED ISSUE #75: BukkitTestingBase supports tear down for each test
Currently, we got BukkitTestingBase for init tests, but since it uses "@BeforeAll" for setting up, this means the data previous testing uses, might leaking to next one and cause testing fails.

It's good to have an option for reload plugins, making clean environment for testing.

image

### PR DESCRIPTION
**Pull request by LeeGodSRC**
There are a LOT of changes here, I will write them all very soon.

### CODE REVIEW

File: buildSrc/build.gradle [hunk 1]
```diff
@@ -20,18 +20,18 @@ gradlePlugin {
...
+    annotationProcessor "org.jetbrains:annotations:23.0.0"
 
     implementation "com.google.code.gson:gson:2.8.6"
-    implementation 'com.google.guava:guava:31.0.1-jre'
-    implementation 'org.apache.commons:commons-lang3:3.12.0'
-    implementation 'org.ow2.asm:asm:9.2'
-    implementation 'org.ow2.asm:asm-commons:9.2'
+    implementation "com.google.guava:guava:31.0.1-jre"
+    implementation "org.apache.commons:commons-lang3:3.12.0"
+    implementation "org.ow2.asm:asm:9.2"
```
Comment (terminalsin) [inline buildSrc/build.gradle]: ASM should be upgraded to 9.3 to support JDK 17+

Comment (terminalsin) [reply buildSrc/build.gradle]: I would recommend having a gradle.settings file with all the versions as opposed to having them all over the project

File: gradle-plugin/build.gradle [hunk 2]
```diff
@@ -40,32 +40,29 @@ pluginBundle {
...
+    compileOnly libs.annotations
+    annotationProcessor libs.annotations
 
-    implementation "com.google.code.gson:gson:" + findProperty("gson.version")
-    implementation 'com.google.guava:guava:31.0.1-jre'
-    implementation 'org.apache.commons:commons-lang3:3.12.0'
+    implementation libs.gson
+    implementation libs.guava
+    implementation libs.common.lang3
     implementation 'org.ow2.asm:asm:9.2'
```
Comment (terminalsin) [inline gradle-plugin/build.gradle]: Same goes here. ASM needs a version upgrade

File: gradle-plugin/src/main/java/io/fairyproject/gradle/util/MavenUtil.java [hunk 3]
```diff
@@ -35,55 +99,124 @@ public String getLatest(String module) throws IOException {
...
-                    final JsonObject itemObject = items.getAsJsonObject();
-
-                    return itemObject.getAsJsonObject("maven2").get("version").getAsString();
+    public String getLatest(String group, String artifact) {
+        return cacheable("getLatest", group + "|" + artifact, String.class, () -> {
+            final URL url = new URL(VERSION_URL + group.replace(".", "/") + "/" + artifact);
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+            connection.setDoInput(true);
+            connection.setRequestMethod("GET");
+            connection.addRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:56.0) Gecko/20100101 Firefox/56.0");
```
Comment (terminalsin) [inline gradle-plugin/src/main/java/io/fairyproject/gradle/util/MavenUtil.java]: User agent should be a custom one if you want to do analytics tracking

**Review by terminalsin — COMMENTED** [review]
Still a WIP review since I didn't have time to go over everything; Here's my current comments:

- Love the new outline and features. Seems great ALTHOUGH I am really not a fan of @Obj. I personally feel @Component is more syntaxically close to Spring, which is what we're inspired one and what we should continue to follow.
- New map bukkit thing looks awesome
- Perhaps it'd be time to think about a new command system based on Brigadier? I can submit an issue with the proposed changes

File: buildSrc/build.gradle (hunk 1 above)
Comment (LeeGodSRC) [reply buildSrc/build.gradle]: done

File: gradle-plugin/src/main/java/io/fairyproject/gradle/util/MavenUtil.java (hunk 3 above)
Comment (LeeGodSRC) [reply gradle-plugin/src/main/java/io/fairyproject/gradle/util/MavenUtil.java]: done

File: gradle-plugin/build.gradle (hunk 2 above)
Comment (LeeGodSRC) [reply gradle-plugin/build.gradle]: done

**Comment by LeeGodSRC** [comment]
The reason I decide to use `@Obj` instead of `@Component` is because how Component has conflict with adventure Component class, which cause awful coding experience when you are using both of them in a single class.
and the new Obj rework has made everything much more flexible. syntaxically it's moving away from spring but in structure it is much more closer

**Comment by LeeGodSRC** [comment]
for Brigadier I don't really have an idea on how the structure is like, you can made an issue about it but I would like to accept a PR about rather than implementing it myself

**Comment by terminalsin** [comment]
Also will player states be something that will be shipped with 0.6?

**Review by terminalsin — APPROVED** [review]
LGTM, make sure to read my suggestion above

**Comment by LeeGodSRC** [comment]
> Also will player states be something that will be shipped with 0.6?

Currently I'm still thinking how player state are going behave under the current structure of fairy, I'll make an issue before working it tho!

**Comment by terminalsin** [comment]
> > Also will player states be something that will be shipped with 0.6?
> 
> Currently I'm still thinking how player state are going behave under the current structure of fairy, I'll make an issue before working it tho!

There's a really simple approach to it. Each MCPlayer data contains the PlayerState. When the setPlayerState method is called, the user's inventory, experience, health, etc... are persisted to storage with the state's instance. At that point the loaded state then will load that same data (unless they override the shouldPersist() method in the PlayerState class) and then call the onLoad function of the state. At that point, the previous state closes all of its listeners whilst the new state registers them and binds any obj adequately.

This effectively allows users to create fast system between swapping from editor mode, mod mode, spectator mode, survival mode, player mode etc...

This furthermore allows for the creation of player traits and so and forth. It isn't too abstract, and isn't too robust. It allows enough wiggleroom to mess plenty with.

**Comment by LeeGodSRC** [comment]
> > > Also will player states be something that will be shipped with 0.6?
> > 
> > 
> > Currently I'm still thinking how player state are going behave under the current structure of fairy, I'll make an issue before working it tho!
> 
> There's a really simple approach to it. Each MCPlayer data contains the PlayerState. When the setPlayerState method is called, the user's inventory, experience, health, etc... are persisted to storage with the state's instance. At that point the loaded state then will load that same data (unless they override the shouldPersist() method in the PlayerState class) and then call the onLoad function of the state. At that point, the previous state closes all of its listeners whilst the new state registers them and binds any obj adequately.
> 
> This effectively allows users to create fast system between swapping from editor mode, mod mode, spectator mode, survival mode, player mode etc...
> 
> This furthermore allows for the creation of player traits and so and forth. It isn't too abstract, and isn't too robust. It allows enough wiggleroom to mess plenty with.

Sounds right, can you write them down in #60 would be helpful!

**Pull request merged**